### ISSUE TITLE
"message": "JWT validation failed: BAD_FORMAT",

### ISSUE THREAD

**Issue by orhan-swe** [cloudendpoints/esp#310]
I am getting following from esp (cloud endpoint with app engine flex):
```
{
 "code": 16,
 "message": "JWT validation failed: BAD_FORMAT",
 "details": [
  {
   "@type": "type.googleapis.com/google.rpc.DebugInfo",
   "stackEntries": [],
   "detail": "auth"
  }
 ]
}
```
I am signing the token with a json key_file (see attached image about the signature being valid)

header:
```
{
  "typ": "JWT",
  "alg": "RS256",
  "kid": "5175a59697375791842e65b064dce1c100d0455a"
}
```
payload
```
{
  "iat": 1510841801,
  "exp": 1510845401,
  "aud": "https://airport5-185713.appspot.com",
  "iss": "airport2-185713@appspot.gserviceaccount.com"
}
```

the actual token:
eyJ0eXAiOiAiSldUIiwgImFsZyI6ICJSUzI1NiIsICJraWQiOiAiNzgzOTY4YmE4Yzc1NWQzMGQ5YTAyYjQ4MmNmMGU3NDJhMzZhNTRlMSJ9.eyJpYXQiOiAxNTEwODQzNTI1LCAiZXhwIjogMTUxMDkxNTUyNSwgImF1ZCI6ICJodHRwczovL2FpcnBvcnQ1LTE4NTcxMy5hcHBzcG90LmNvbSIsICJpc3MiOiAiY2xpZW50MUBhaXJwb3J0NS0xODU3MTMuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20ifQ==.BG_4vPJBzi_T5ZkBwtgx33553L3wSy9QeDxDWaEWvJAZEg4l0XZqclBL_p7PzFAjYhkQrmAY4CCyFqQ_ZCAM78mqIzbmy4uduyuyqm12miclJV62edNPSMXUjyTKk5C4pGsJ4fr02uf6BGENrMlS7COnXXUfS-kBIjxHmErBL49ItH8gYB2xJ30iON5W_HuqfH1QR_vch9-DWMcg8E2iSCuRR5GUBrR9kww35E4nhdFcvRYIA2eX4F4zPX5DEpPWsSHfE-TvPZIAbx1r_-PPkkHfwMfjXoatEfqw8m2EFfaKMDxxKZi_26iAQ9MQ1TqeFfGfLGhI4xZ0fwx8ZPfaGA==

certificate public key (783968ba8c755d30d9a02b482cf0e742a36a54e1):
[https://www.googleapis.com/robot/v1/metadata/x509/client1@airport5-185713.iam.gserviceaccount.com]

openAPI.yaml:
```
google_jwt_client-2:
    authorizationUrl: ""
    flow: "implicit"
    type: "oauth2"
    x-google-issuer: "client1@airport5-185713.iam.gserviceaccount.com"
    x-google-jwks_uri: "https://www.googleapis.com/robot/v1/metadata/x509/client1@airport5-185713.iam.gserviceaccount.com"
    x-google-audiences: "https://airport5-185713.appspot.com"
```

I am generating the token with:
```
def generate_jwt(service_account_file = file_path):

        signer = google.auth.crypt.RSASigner.from_service_account_file(
            service_account_file)
        
        now = int(time.time())
        expires = now + 3600*20  # twenty hours

        payload = {
            'iat': now,
            'exp': expires,
            'aud': 'https://airport5-185713.appspot.com',
            'iss': 'client1@airport5-185713.iam.gserviceaccount.com'
        }

        jwt = google.auth.jwt.encode(signer, payload)
        logging.debug(jwt)

        return jwt
```

to reproduce: 
curl -H "Authorization: Bearer ${TOKEN}" "${ENDPOINTS_HOST}/airportName"
where  ENDPOINTS_HOST=https://airport5-185713.appspot.com

![image](https://user-images.githubusercontent.com/10519489/32897649-7d4f480a-cae6-11e7-961a-2ce67f42353d.png)

**Comment by qiwzhang** [comment]
@liminw could you help?

**Comment by liminw** [comment]
I think the problem here is, the "sub" claim is missing in the token. 

@orhan-swe , could you please add "sub" claim to the token and try it again?

**Comment by orhan-swe** [comment]
@liminw  Thanks, but it did not help, I am getting the same response

generated token:
eyJ0eXAiOiAiSldUIiwgImFsZyI6ICJSUzI1NiIsICJraWQiOiAiNzgzOTY4YmE4Yzc1NWQzMGQ5YTAyYjQ4MmNmMGU3NDJhMzZhNTRlMSJ9.eyJpYXQiOiAxNTEwOTA5MTgyLCAiZXhwIjogMTUxMDk4MTE4MiwgImF1ZCI6ICJodHRwczovL2FpcnBvcnQ1LTE4NTcxMy5hcHBzcG90LmNvbSIsICJpc3MiOiAiY2xpZW50MUBhaXJwb3J0NS0xODU3MTMuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20iLCAic3ViIjogIjExMjc2ODE3Mjg4MTM0OTA3NTEzOSIsICJlbWFpbCI6ICJjbGllbnQxQGFpcnBvcnQ1LTE4NTcxMy5pYW0uZ3NlcnZpY2VhY2NvdW50LmNvbSJ9.PuUi6f9VFxynryVCibQ9pnetvEmBLqn02I7YVJp1sAz0HCC-pqoCFH8sgHbif3soln_jV-w3JyACzCjN8Y9B0QIn-cv3hfzY08p8pckButHFsB-ycZd2ZuEAZhTjNp9lJxm1BdjIsjH1bMzOeu2bhA9JxwJVTHXCHOPi9VQPx_9ylJwOgJycWDcPIRT2A-NtDdH8J9oVSm5oOtls8nOwMNT4sN42u-s66IX7YdZ0OblY7xLFhoqW4aV9PhTDzp9fMqYOKyhCzK0atAwTNSNRNjHWRIJavEEHLr2RzvEmZUAo62fq8tFaN8ul6ngt20lSdGmTE-xsUhsR6Rt0sdIwWQ==

payload = {
        'iat': now,
        'exp': expires,
        'aud': 'https://airport5-185713.appspot.com',
        'iss': 'client1@airport5-185713.iam.gserviceaccount.com',
        'sub': '112768172881349075139',
        'email': 'client1@airport5-185713.iam.gserviceaccount.com'
    }

**Comment by liminw** [comment]
@orhan-swe , I tried the new token you posted, the error becomes "JWT validation failed: UNKNOWN". It is likely caused by the fact that your "sub" and "email" fields are different. Could you try to make "sub" the same as your "iss" claim (client1@airport5-185713.iam.gserviceaccount.com)? 

There is implicit requirement that if the "iss" (issuer) claim is an email address, then the "sub" (subject) and "iss" claims should be the same. This is to ensure that for e-mail issuers, the JWT is self issued. I understand the error message ("UNKNOWN") is not helpful here. We documented it on https://cloud.google.com/endpoints/docs/openapi/troubleshoot-jwt-openapi.

**Comment by orhan-swe** [comment]
@liminw i generated and tried a new token after i wrote the note above, it may have made the old one invalid and you are getting unknown; I am not sure if I ever tried with both iss and sub being same email address, though I did try many different combinations. I will try this on Monday and see if it gives different result. Thanks

**Comment by orhan-swe** [comment]
@liminw 
The issue has been resolved. 

It seems that I had two different issues.
The main issue was that the encode function below was giving a string bytes object and not a normal string: 
`jwt = google.auth.jwt.encode(signer, payload)`

I was using the python sample project from here:
https://cloud.google.com/endpoints/docs/openapi/service-to-service-auth#making_an_authenticated_call_to_an_endpoints_api

A note should be added about this in documentation, I got the sample working by changing the line to this:
`jwt = google.auth.jwt.encode(signer, payload).decode('UTF-8')`

The other problem is that "sub" is required and it must be same as 'iss' in this case just like you said. Thank you very much all the help!

To summarize: 

- sending byte encoded string would result in "BAD_FORMAT"
- sending without the "iss" would result in "BAD_FORMAT"
- sending with "iss" not identical to sub would result in "UNKNOWN"

As you can see the two different issues ware giving me same "BAD_FORMAT" message. This made it more difficult to debug.

This issue can be closed.

**Comment by qiwzhang** [comment]
@orhan-swe Thanks for your detail info.  We certainly don't want other users to run into this issue again.  @liminw please use this issue to update our doc.   Thanks.

**Comment by liminw** [comment]
@orhan-swe Thanks for the info! We will update the doc accordingly.
